### ISSUE TITLE
XHR does not treat 4xx responses as errors

### ISSUE THREAD

**Issue by escardin** [angular/angular#5130]
When my server returns a 400 response with text, the xhr backend does not route this through to the error observer, and instead calls the success path.

I am porting a service from angular 1 to angular 2, and this was not a problem in angular 1, nor is it a problem with rxjs-dom, which I am using in the meantime.

Affected area:
https://github.com/angular/angular/blob/master/modules/angular2/src/http/backends/xhr_backend.ts#L41

What rxjs-dom does and I think angular 2 should do similarly:
https://github.com/Reactive-Extensions/RxJS-DOM/blob/master/src/ajax/ajax.js#L76

**Comment by jeffbcross** [comment]
Agreed. I thought there was already an issue for this, but hard to search on my mobile.

**Comment by jeffbcross** [comment]
Though some have disagreed with this behavior because fetch only errors on network errors.

**Comment by escardin** [comment]
I was thinking about that when I filed this issue. If it was a promise, it would be success and error, and http codes only consider 2xx to be success. Since subscribe is analogous to success, it should only be 2xx codes.

Realistically, anything other than 2xx is an error (Or rather, not what I am expecting), and needs special handling, the same sort of handling I am doing when I have a network error.

**Comment by robwormald** [comment]
Just depends really if we want to stick to the semantics of the fetch specification or not. I'm mostly in agreement here, though perhaps it should be a configurable/injectable option.

Proposed handler from [github's fetch polyfill](https://github.com/github/fetch#handling-http-error-statuses)
```
function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response
  } else {
    var error = new Error(response.statusText)
    error.response = response
    throw error
  }
}
```

**Comment by e-oz** [comment]
it's really awful, especially because in case of error we can't read error status. Please please fix it.

**Comment by danielgerlag** [comment]
I don't agree with this.  a non 200 response is not an error, can we revert this back to the way it was?

**Comment by escardin** [comment]
4xx and 5xx are by definition, errors. Says right here: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

10.4 Client Error 4xx

10.5 Server Error 5xx

They're admittedly different from a lost connection, but they're still errors, handled differently from your 'happy path'. Handling of a lost connection and a server or client error are more likely to be similar than a server/client error and a successful response.

If this is reverted, then everyone has to check status codes on their responses for the happy path, just to be sure it's right, rather than that path having a clear route that is predictable and handling special cases with an error handler.

You can work around this by catching the error and transforming it into a 'success' if that's how you want your code to work.

**Comment by danielgerlag** [comment]
a 4xx could be a validation error with the data submitted, this is not an exception, with ASP.NET WebAPI you would also get a ModelState in the body content with the 400 response, now I can't access that anymore.  Only 5xx responses are truly exceptions. A 4xx response means the communication was successful but the server didn't like the data it got, still a success.

**Comment by danielgerlag** [comment]
We should revert the default backend to not throw this error and create an alternate ConnectionBackend for those who want to treat non 200 responses as actual errors.

**Comment by e-oz** [comment]
no

**Comment by e-oz** [comment]
I like when people decide what other people should do. So funny.

**Comment by danielgerlag** [comment]
There has been no due process of consensus on this issue.  If this is the kind of governance on this project, I may have to reconsider using ng2 for any serious projects.

**Comment by escardin** [comment]
'a 4xx could be a validation error' is still an error. Just because you get content in your 4xx response doesn't mean everyone else does, and is as far as I know, more the exception than the rule. You can catch the error, check if it's 4xx and redirect it through your normal pipe. You could make a wrapper around http and make that the default. You could extend XHR backend and change it back to how you like. There are lots of solutions.

Neither e-oz nor myself are members of the Angular 2 team, but we like it how it is.

I do have to say though, this isn't a democracy, due process is not strictly a consideration. You've made your case, and we are disagreeing with you. If you want consensus, anecdotally, I've seen more complaints about it doing what you ask than doing what I asked.

**Comment by danielgerlag** [comment]
How does one flag this issue so that someone from the core angular team can review it?

**Comment by escardin** [comment]
This would do exactly what you want: http://plnkr.co/edit/dsNKsUImdbh4uOvaS1kl?p=preview

**Comment by danielgerlag** [comment]
No, that is a hack.  The way the ConnectionBackend originally was the most pure and gave the developer the most flexibility.  It should not sniff the contents of the response to decide if an error occurred, it should only care that it got or did not get a response from the server.  It's up the the consumer to make decisions based on the contents of the response from the server.

**Comment by robwormald** [comment]
@danielgerlag I, in fact, agree with you that the ConnectionBackend itself shouldn't validate status codes, and only error on network failures - this is closer to how the fetch specification handles it, and it's likely I'll make that change and have Http itself do the status check.

That said, I'd encourage *everybody* in this thread to treat each other with civility and respect, per our code of conduct. Angular2 is evolving, and a little bit of patience goes a long way.

**Comment by escardin** [comment]
I'm happy with that solution.

**Comment by danielgerlag** [comment]
What do you mean be "Http itself do the status check"?

**Comment by escardin** [comment]
@danielgerlag I'm pretty sure with it as it is now, you can still detect a connection failure. It is hard to come up with a sample proving it though. The best I could come up with was a gibberish url that got blocked by cors and gave a 200 response.

**Comment by robwormald** [comment]
> What do you mean be "Http itself do the status check"?

XHRBackend will only error on network failures. Http will do a (configurable) status code check, similar to what the backend is currently doing. This is a happy middle path between the fetch-like low level behavior while maintaining the ng1 / jqXHR behavior of erroring on > 299 errors

**Comment by danielgerlag** [comment]
Awesome, thank you

**Comment by sinedied** [comment]
Has anything changed related to this? I'm having this exact issue while from what I understand it should have been fixed?

When the server returns a 404 error, it does not get through the error handler, but the "standard" handler.

I'm using `@angular/http 2.4.8`.

**Comment by DzmitryShylovich** [comment]
> Has anything changed related to this?

nope https://github.com/angular/angular/blob/master/modules/%40angular/http/src/http_utils.ts#L33

**Comment by sinedied** [comment]
I found the issue: the `MockBackend` used for unit testing does not follow the same behavior as the `XHRBackend` used by the app. Should I raise a separate issue?

**Comment by DzmitryShylovich** [comment]
@sinedied duplicate of https://github.com/angular/angular/issues/12857#issuecomment-266490184 ?

**Comment by sinedied** [comment]
Oups, my bad!
